go/vt/vttablet/tabletmanager/rpc_vreplication.go

```diff
@@ -139,6 +143,85 @@ func (tm *TabletManager) CreateVReplicationWorkflow(ctx context.Context, req *ta
 	return &tabletmanagerdatapb.CreateVReplicationWorkflowResponse{Result: sqltypes.ResultToProto3(res)}, nil
 }
 
+// DeleteTableData will delete data from the given tables (keys in the
+// req.Tabletfilters map) using the given filter or WHERE clauses (values
+// in the map). It will perform this work in batches of req.BatchSize
+// until all matching rows have been deleted in all tables, or the context
+// expires.
+func (tm *TabletManager) DeleteTableData(ctx context.Context, req *tabletmanagerdatapb.DeleteTableDataRequest) (*tabletmanagerdatapb.DeleteTableDataResponse, error) {
+	if req == nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid nil request")
+	}
+
+	if len(req.TableFilters) == 0 { // Nothing to do
+		return &tabletmanagerdatapb.DeleteTableDataResponse{}, nil
+	}
+
+	// So that we do them in a predictable and uniform order.
+	tables := maps.Keys(req.TableFilters)
+	sort.Strings(tables)
+
+	batchSize := req.BatchSize
+	if batchSize < 1 {
+		batchSize = movetables.DefaultDeleteBatchSize
+	}
+	limit := &sqlparser.Limit{Rowcount: sqlparser.NewIntLiteral(fmt.Sprintf("%d", batchSize))}
+
+	throttledLogger := logutil.NewThrottledLogger("DeleteTableData", 1*time.Minute)
+	checkIfCanceled := func() error {
+		select {
+		case <-ctx.Done():
+			return vterrors.Wrap(ctx.Err(), "context expired while deleting data")
+		default:
+			return nil
+		}
+	}
+
+	for _, table := range tables {
+		stmt, err := tm.Env.Parser().Parse(fmt.Sprintf("delete from %s %s", table, req.TableFilters[table]))
+		if err != nil {
+			return nil, vterrors.Wrapf(err, "unable to build delete query for table %s", table)
+		}
+		del, ok := stmt.(*sqlparser.Delete)
+		if !ok {
+			return nil, vterrors.Wrapf(err, "unable to build delete query for table %s", table)
+		}
+		del.Limit = limit
+		query := sqlparser.String(del)
+		// Delete all of the matching rows from the table, in batches, until we've
+		// deleted them all.
+		for {
```

Comment:
I am wondering,  for debugging/monitoring purposes, if we should temporarily log periodically (say every 100K rows) while deleting and/or at the start/end of a table, so that we can track any issues. We can remove it once the functionality is proven if we see that logs are spammed or leave it in.

Since Cancel will be rarely called and we would probably want to monitor performance impact when that happens and correlate with the logging.

Reply:
I did this here: https://github.com/vitessio/vitess/pull/16906/commits/e7374c8575f5f0552034050c9980ad62b21ab568

I think that's a good minimum, and we can add additional debug logging if needed. I don't personally think it's worth logging more here w/o good reason as we can get information from MySQL about how many times the delete has run, how long it took, how long it is between each delete (because we're throttled), etc from the binary logs, performance_schema, etc.

Reply:
After thinking about it a bit more, some additional progress info that can be easily seen in the vttablet logs can't really hurt and the overhead is very low so I added something here: https://github.com/vitessio/vitess/pull/16906/commits/a993ce2577452472a8b4af2ddaa9fb0c9fe1fc9b

Reply:
So... I changed my mind and decided the bite the bullet as I had the problem space clear in my head ATM 😄 https://github.com/vitessio/vitess/pull/16906/commits/6a4c79e69a2125b9467fb8d7b176c3d30152a92f